src/functions/cmp.ts

```diff
@@ -1,18 +1,89 @@
-import { domDeps } from '../classes/dexie/dexie-dom-dependencies';
-import { exceptions } from '../errors';
-import { IndexableType } from '../public';
+// Implementation of https://www.w3.org/TR/IndexedDB-3/#compare-two-keys
 
-let _cmp: (a: any, b: any) => number;
-export function cmp(a: IndexableType, b: IndexableType): number {
-  if (_cmp) return _cmp(a, b);
-  const {indexedDB} = domDeps;
-  if (!indexedDB) throw new exceptions.MissingAPI();
-  _cmp = (a, b) => {
-    try {
-      return a == null || b == null ? NaN : indexedDB.cmp(a, b);
-    } catch {
-      return NaN;
+import { toStringTag } from './utils';
+
+// ... with the adjustment to return NaN instead of throwing.
+export function cmp(a: any, b: any): number {
+  try {
+    const ta = type(a);
+    const tb = type(b);
+    if (ta !== tb) {
+      if (ta === 'Array') return 1;
+      if (tb === 'Array') return -1;
+      if (ta === 'binary') return 1;
+      if (tb === 'binary') return -1;
+      if (ta === 'string') return 1;
+      if (tb === 'string') return -1;
+      if (ta === 'Date') return 1;
+      if (tb !== 'Date') return NaN;
+      return -1;
+    }
+    switch (ta) {
+      case 'number':
+      case 'Date':
+      case 'string':
+        return a > b ? 1 : a < b ? -1 : 0;
+      case 'binary': {
+        return compareUint8Arrays(getUint8Array(a), getUint8Array(b));
+      }
+      case 'Array':
+        return compareArrays(a, b);
     }
+  } catch {}
+  return NaN; // Return value if any given args are valid keys.
+}
+
+export function compareArrays(a: any[], b: any[]): number {
+  const al = a.length;
+  const bl = b.length;
+  const l = al < bl ? al : bl;
+  for (let i = 0; i < l; ++i) {
+    const res = cmp(a[i], b[i]);
+    if (res !== 0) return res;
+  }
+  return al === bl ? 0 : al < bl ? -1 : 1;
+}
+
+export function compareUint8Arrays(
+  a: Uint8Array,
+  b: Uint8Array
+) {
+  const al = a.length;
+  const bl = b.length;
+  const l = al < bl ? al : bl;
+  for (let i = 0; i < l; ++i) {
+    if (a[i] !== b[i]) return a[i] < b[i] ? -1 : 1;
```

Comment:
...as a follow up to my previous comment, here's an experiment with `DataView`

```js
function compareUint8Arrays(a,b) {
    const dv1 = new DataView(a.buffer);
    const dv2 = new DataView(b.buffer);
    let idx = 0;
    let bytes = Math.min(dv1.byteLength, dv2.byteLength);
    if (bytes > 4) {
        let len = bytes >> 2;
        while (idx < len) {
            let b1 = dv1.getUint32(idx);
            let b2 = dv2.getUint32(idx);
            if (b1 !== b2) return b1 < b2 ? -1 : 1;
            ++idx;
        }
        bytes -= idx << 2;
    }
    if (bytes > 2) {
        let len = bytes >> 1;
        while (idx < len) {
            let b1 = dv1.getUint16(idx);
            let b2 = dv2.getUint16(idx);
            if (b1 !== b2) return b1 < b2 ? -1 : 1;
            ++idx;
        }
        bytes -= idx << 1;
    }
    while (idx < bytes) {
        let b1 = dv1.getUint8(idx);
        let b2 = dv2.getUint8(idx);
        if (b1 !== b2) return b1 < b2 ? -1 : 1;
        ++idx;
    }
    return dv1.byteLength === dv2.byteLength ? 0 : dv1.byteLength < dv2.byteLength ? -1 : 1;
}
```

```js
b1=cryptHelper.getRandomValues(4730)
b2=b1

> console.time('dd');for(let i = 1e6; i--; a2());console.timeEnd('dd');
< dd: 1382.230224609375 ms
```

... 83% faster with 4KB of data (?) but still slower than `indexedDB.cmp`

Reply:
If I get it right, the conclusion of all this is that:

* JS based comparison still beats native one by far for numbers and strings (and probably Dates and reasonably short arrays)
* The only situation where we can see native indexedDB.cmp beat the JS based one is when comparing large binaries that have an identical start sequence in their first 500 bytes.
* Binary keys still perform much better with the JS based algorithm than the native one as long as they are reasonably short or have reasonable randomness.

Now, when we are comparing keys, we only do it for indexed keys and properties and these shall not be large anyway (see [#1392](https://github.com/dexie/Dexie.js/issues/1392#issuecomment-927841921)) and there is almost never any good reason to index large binaries. They typically binary index represent either a primary key of maybe 16 (random) bytes, or digests that may be between 8-32 bytes and have the same randomness. There might be other use cases, like indexing paths of tree structures that may be longer, and having identical start sequences, but probably never that large. Indexing binaries larger than say 1500 bytes would create other performance issues anyway because it bloats the BTree in the database - which is the reason that most databases limit the permitted size of BTree indexes to 1700-3500 bytes depending on DB engine.